Feature Folder and Multi branch Workflow Support

zack-hable opened the pull request:
This PR should resolve the following issues:

- #11 Multi Branch Pipelines have been added as 'folders'
- #9 Folders have been added and their nested jobs
- #18 Use of an absolute job path or job id to build job

Changes:
- Added a JenkinsJob and JenkinsFolder class to better match the structure of Jenkins Objects.
- Modified List method to load information about the jobs in a folder
- Modified getJob() method to figure out if an getById method was called (load our stored job object) or if the user only gave us a jobName or a path and jobName to search by.

Hopefully @wintondeshong or others could give this a look/test.

wintondeshong commented:
@zack-hable Thanks for all the work on this PR! I'm going to give the code an initial review now and hopefully give it a test today or tomorrow.

wintondeshong commented on src/jenkins-enhanced.coffee:
```diff
@@ -353,13 +486,86 @@ class HubotJenkinsPlugin extends HubotMessenger
-    job     = aliases[job] if aliases[job]
-
-    if escape then @_querystring.escape(job) else job
+    # check if its a job we already stored
```
Consider breaking the `_getJob` method into separate smaller methods with descriptive names to reduce the nesting of these if/else statements.

wintondeshong commented on src/jenkins-enhanced.coffee:
```diff
@@ -353,13 +486,86 @@ class HubotJenkinsPlugin extends HubotMessenger
+    if (typeof(@msg.match[1]) == "object")
+      return @msg.match[1]
+    # check if the user gave us a folder path to follow and a job
+    else
```
Being the initial `if` expression does a hard return, this `else` can be removed and code following can be unindented to reduce nesting.

wintondeshong commented on src/jenkins-enhanced.coffee:
```diff
@@ -353,13 +486,86 @@ class HubotJenkinsPlugin extends HubotMessenger
+            @send "There are no folders with the name #{folderPath[0]}."
+			
+        else
+          # this is for an absolute path to the folder
```
Perhaps take the body of this `else` and wrap into method named `_getJobByAbsolutePath` (or something like that)

wintondeshong commented on src/jenkins-enhanced.coffee:
```diff
@@ -353,13 +486,86 @@ class HubotJenkinsPlugin extends HubotMessenger
+          else
+            @send "There are no folders with the path #{folderPath.join('/')}."
+      else
+        # this is when they just give us a job name
```
Perhaps take the body of this else and wrap into method named `_getJobByName` (or something like that)

zack-hable commented:
@wintondeshong thanks for the feedback!  I'll try to incorporate those changes and push another commit.

zack-hable commented:
I've split the _getJob() method up and added support for referencing the root folder for the <job folder/job> parameter.  For example you would use `jenkins describe /my root job` as there is a possibility that a nested job my also have the same name.

wintondeshong commented:
@zack-hable Here are my findings after testing various commands:

**Question**
When attempting to issue a 'build' command of `jenkins build TestFolder/TestJob` I get the message 'There are no jobs with the name TestFolder'. If I only use the job name of 'TestJob', it works just fine, but when attempting to provide folder path it fails. Am I missing something? 

**Nice to haves** (won't hold back acceptance of pull request)
servers: Would be great to have full paths of job names listed. In the event a job has the same name across folders and root, you can't distinguish them.

**Verified**
list: Dig the hierarchy indentation of the list view 👍 

**Testing Details**
Jenkins: v2.121.2
OS: Windows

zack-hable commented:
hey @wintondeshong ,

I've changed the regex of the build command to match those of describe/last build and you should now be able to build with folder support.  I've also updated the servers command to show the nested folders and jobs.  Thanks for the feedback and testing!  Hopefully everything should be resolved now.

wintondeshong commented:
@zack-hable thanks for all the work on this pull request. Gave it all another run through and works as expected! Nice work!
